core/runtime/binaryen/runtime_manager.cpp

```diff
@@ -30,18 +31,51 @@ namespace kagome::runtime::binaryen {
   RuntimeManager::RuntimeManager(
       std::shared_ptr<runtime::WasmProvider> wasm_provider,
       std::shared_ptr<extensions::ExtensionFactory> extension_factory,
+      std::shared_ptr<storage::trie::TrieStorage> trie_storage,
       std::shared_ptr<crypto::Hasher> hasher)
-      : wasm_provider_(std::move(wasm_provider)),
-        extension_factory_(std::move(extension_factory)),
-        hasher_(std::move(hasher)) {
+      : wasm_provider_{std::move(wasm_provider)},
+        storage_{std::move(trie_storage)},
+        extension_factory_{std::move(extension_factory)},
+        hasher_{std::move(hasher)} {
     BOOST_ASSERT(wasm_provider_);
+    BOOST_ASSERT(storage_);
     BOOST_ASSERT(extension_factory_);
     BOOST_ASSERT(hasher_);
   }
 
-  outcome::result<std::tuple<std::shared_ptr<wasm::ModuleInstance>,
-                             std::shared_ptr<WasmMemory>>>
-  RuntimeManager::getRuntimeEnvironment() {
+  outcome::result<RuntimeManager::RuntimeEnvironment>
+  RuntimeManager::createPersistentRuntimeEnvironmentAt(
+      const common::Hash256 &state_root) {
+    auto persistent_batch = storage_->getPersistentBatchAt(state_root).value();
+    persistent_batch_ = std::move(persistent_batch);
+    return createRuntimeEnvironment(persistent_batch_);
+  }
+
+  outcome::result<RuntimeManager::RuntimeEnvironment>
+  RuntimeManager::createEphemeralRuntimeEnvironmentAt(
```

Comment:
What reason to make separate methods by state_root existence? In more places we should check his existence, and handle it on diffeternt ways. I suggest combine it, for avoiding if-branches in more places, instead single in combined method.

Reply:
Because as for me the code is a bit more self-documented this way. Different situations require either applying methods on the current root, or switch to a particular root. As for current usage it is decided which one to use in a single combined method in **runtime_api.hpp**. If to refactor it to two methods with optional argument, one conditional operator would just move inside every method from **runtime_api.hpp**, so I don't see much difference in terms of num of branches.